sunpy/map/sources/asos.py

```diff
@@ -0,0 +1,178 @@
+        according to the 'date_start' with sunpy.coordinates.get_earth
+        
+        Usage: set keyword 'fix_hxi_observer' when using sunpy.map.Map(), default:True.
+               if not fix, the sunpy will fix it automatically in some cases.
+        """
+        if not (set(['HGLN_OBS','HGLT_OBS','DSUN_OBS']) < set(self.fits_header)):
+            
+            from sunpy.coordinates import get_earth
+            observer = get_earth(self.date_start)
+            
+            if observer.frame.name=='heliographic_stonyhurst':
+                self.meta['HGLN_OBS'] = observer.lon.to_value(u.deg)
+                self.meta['HGLT_OBS'] = observer.lat.to_value(u.deg)
+                self.meta['DSUN_OBS'] = observer.radius.to_value(u.m)
+                #log.info("Assuming Earth-based observer to set dsun_obs,hgln_obs,hglt_obs.")
+            else:
+                raise ValueError("observer.frame.name not eq 'heliographic_stonyhurst', not support")
+
+    def _get_cmap_name(self):
+        """
+        Using default cmap the same as RHESSI
+        """
+        return "rhessi" 
+
+    def _rotation_matrix_from_crota(self):
+        """
+        Rotation in CROTA. HXI may not need
+        """
+        return super()._rotation_matrix_from_crota(crota_key='CROTA')
+
+    @property
+    def spatial_units(self):
+        """
+        If cunit{1 or 2} are equal to 'arcsecs', assumes that cunit{1 or 2}
+        respectively are intended to be 'arcsec'.
+        """
+        units = [self.meta.get('cunit1', None), self.meta.get('cunit2', None)]
+        if self.meta['cunit1'] == 'arcsecs':
+            units[0] = 'arcsec'
+        if self.meta['cunit2'] == 'arcsecs':
+            units[1] = 'arcsec'
+        units = [None if unit is None else u.Unit(unit.lower()) for unit in units]
+        return SpatialPair(units[0], units[1])
+
+    @property
+    def coordinate_system(self):
+        """
+        If CTYPE{1 or 2} are equal to 'solar_x' or 'solar_y', assumes that CTYPE{1 or 2}
+        respectively are intended to be 'HPLN-TAN' or 'HPLT-TAN'.
+        """
+        ctypes = [self.meta['ctype1'], self.meta['ctype2']]
+        if ctypes[0] == 'solar_x':
+            ctypes[0] = 'HPLN-TAN'
+        if ctypes[1] == 'solar_y':
+            ctypes[1] = 'HPLT-TAN'
+        return SpatialPair(*ctypes)
+    
+    @property
+    def date_start(self):
+        """DATE-OBS is the start date"""
+        date_obs = self.meta.get('DATE-OBS')
+        if len(date_obs.split(" ")[0])==9:
+            date_obs = date_obs[:7]+'20'+date_obs[7:]
+        return parse_time(date_obs)
+
+    @property
+    def date_end(self):
+        return self.date_start + self.meta.get('exptime')*u.second
+    
+    @property
+    def _date_obs(self):
+        """using date_start"""
+        return self.date_start
+        
+    @property
+    def reference_date(self):
+        """using date_start, compatible with the old version of sunpy """
```

Comment:
What old version of sunpy?

Reply:
I see, I would only consider the oldest version being 7.0, no need to check what happens on 6.X.